Add support for EnvironmentOption to pax-exam-container-forked

glimmerveen opened the pull request:
It is a slight variation on how the KarafJavaRunner interprets this option. It allows for:
* Passthrough of an externally defined variable: new EnvironmentOption("HOME")
* Setting the environment variable: new EnvironmentOption("VAR=value") 

The latter can of course also be used to redefine an externally existing environment variable.

glimmerveen commented:
@oliverlietz any thoughts on this proposed extension to allow the use of the EnvironmentOption within the forked container?

glimmerveen commented:
> @oliverlietz any thoughts on this proposed extension to allow the use of the EnvironmentOption within the forked container?

glimmerveen commented:
@oliverlietz could have a look at the proposed extension? I have updated the branch to resolve the conflicts that were there.

oliverlietz commented on containers/pax-exam-container-forked/pom.xml:
```diff
@@ -172,6 +172,21 @@
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <environmentVariables>
```
Make use of [System Stubs](https://github.com/webcompere/system-stubs) to test with environment variables.

oliverlietz commented on core/pax-exam/src/main/java/org/ops4j/pax/exam/ExamJavaRunner.java:
```diff
@@ -134,13 +136,59 @@ public synchronized void exec(final String[] vmOptions,
+            } else {
+                key = envOption.substring(0, equalsPosition).trim();
+                value = envOption.substring(equalsPosition + 1).trim();
+            }
```
Let's not modify (trim) input and keep it as is.

oliverlietz commented on core/pax-exam/src/main/java/org/ops4j/pax/exam/ExamJavaRunner.java:
```diff
@@ -134,13 +136,59 @@ public synchronized void exec(final String[] vmOptions,
+            }
+            if (envMap.containsKey(key)) {
+                throw new IllegalArgumentException("Duplicate key found: " + key);
+            }
```
Did you encounter problems in above situations where `IllegalArgumentException`s are thrown? Or would logging a warning be sufficient?

glimmerveen replied:
No I did not encounter issues, I added this as sanity check on the user input. I guess it personal preference to prefer fail fast by throwing an exception, as in my experience warnings are often overlooked. 
Logging a warning is fine by me as well, please let me know if you prefer that.

oliverlietz commented:
@glimmerveen, Please sign your commits with GPG and include the issue number in the subject line of the commit message. Thanks!

glimmerveen commented:
> @glimmerveen, Please sign your commits with GPG and include the issue number in the subject line of the commit message. Thanks!

I have pushed a new GPG signed commit, incorporated the PR feedback (except one item), and referenced the PR number in the commit message.

glimmerveen committed:
This adds support for the EnvironmentOption to pax-exam-container-forked. (#1127)

It is a slight variation on how the KarafJavaRunner interprets this option. It allows for:
* Passthrough of an externally defined variable: new EnvironmentOption("HOME")
* Setting the environment variable: new EnvironmentOption("VAR=value")
The latter can of course also be used to redefine an externally existing environment variable.

oliverlietz commented:
Thanks, @glimmerveen!

oliverlietz commented:
@glimmerveen, Can you reopen the PR? It seems that your change is causing issues on Windows. After merging the change there was no successful Windows build anymore. Failure:

```
Exception in thread "main" java.rmi.server.ExportException: Listen failed on port: 0; nested exception is: 
	java.net.SocketException: Unrecognized Windows Sockets error: 10106: create
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:337)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.exportObject(TCPTransport.java:243)
	at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.exportObject(TCPEndpoint.java:412)
	at java.rmi/sun.rmi.transport.LiveRef.exportObject(LiveRef.java:147)
	at java.rmi/sun.rmi.server.UnicastServerRef.exportObject(UnicastServerRef.java:234)
	at java.rmi/java.rmi.server.UnicastRemoteObject.exportObject(UnicastRemoteObject.java:470)
	at java.rmi/java.rmi.server.UnicastRemoteObject.exportObject(UnicastRemoteObject.java:355)
	at org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl.export(RemoteFrameworkImpl.java:94)
	at org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl.<init>(RemoteFrameworkImpl.java:79)
	at org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl.main(RemoteFrameworkImpl.java:448)
Caused by: java.net.SocketException: Unrecognized Windows Sockets error: 10106: create
	at java.base/java.net.ServerSocket.createImpl(ServerSocket.java:327)
	at java.base/java.net.ServerSocket.getImpl(ServerSocket.java:277)
	at java.base/java.net.ServerSocket.bind(ServerSocket.java:395)
	at java.base/java.net.ServerSocket.<init>(ServerSocket.java:257)
	at java.base/java.net.ServerSocket.<init>(ServerSocket.java:149)
	at java.rmi/sun.rmi.transport.tcp.TCPDirectSocketFactory.createServerSocket(TCPDirectSocketFactory.java:45)
	at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.newServerSocket(TCPEndpoint.java:670)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:324)
	... 9 more
```

I guess a required environment variable `SystemRoot` is no longer available to the forked process.

glimmerveen commented:
@oliverlietz sure, I can look into this, though I don't have access to a Windows machine to perform tests.

At first glance, I am a bit puzzled to relate this to my change. I believe the behaviour before my change was already that all environment variables from the environment that is about the fork a test were stripped, unless explicitly referenced by the 'envOptions' (see https://github.com/ops4j/org.ops4j.pax.exam/blob/4.14.0/core/pax-exam/src/main/java/org/ops4j/pax/exam/ExamJavaRunner.java#L136-L144). The forked factory before my change never even supplied these envOptions, and thus all variables would have been removed, resulting in the forked process running in an empty environment.

Is perhaps the dependency on the SystemRoot environment variable something that has been changed?

glimmerveen commented:
> At first glance, I am a bit puzzled to relate this to my change. I believe the behaviour before my change was already that all environment variables from the environment that is about the fork a test were stripped, unless explicitly referenced by the 'envOptions' (see https://github.com/ops4j/org.ops4j.pax.exam/blob/4.14.0/core/pax-exam/src/main/java/org/ops4j/pax/exam/ExamJavaRunner.java#L136-L144). The forked factory before my change never even supplied these envOptions, and thus all variables would have been removed, resulting in the forked process running in an empty environment.

I have pushed a new branch in which I reverted the commit related to this PR, and did a subsequent commits to print the environment variables within the forked test. On the windows build I do see this being set, with the aforementioned (original) environment setup that I expected to result in an empty environment as well, but apparently it does not. 

What is still puzzling is that I also added 2 log statements within the setEnv function, and as I expected, from the ExamJavaRunner's perspective, and empty environment is used, even though I see SystemRoot being set...

glimmerveen commented:
> I have pushed a new branch in which I reverted the commit related to this PR, and did a subsequent commits to print the environment variables within the forked test. On the windows build I do see this being set, with the aforementioned (original) environment setup that I expected to result in an empty environment as well, but apparently it does not.
> 
> What is still puzzling is that I also added 2 log statements within the setEnv function, and as I expected, from the ExamJavaRunner's perspective, and empty environment is used, even though I see SystemRoot being set...

Upon cherry-picking the change of this PR again, the 2 log statements (before and after the environment setup) output the same, suggesting it does the same thing, however now the tests fail (as it does on v4.x).

I will have another look at the implementation from this PR...

oliverlietz commented:
It took me some time to understand that the failures are not from changes to the Windows runners or networking but from the environment variables handling.
Let's document our findings in source code once it's working properly.

glimmerveen commented:
> It took me some time to understand that the failures are not from changes to the Windows runners or networking but from the environment variables handling. Let's document our findings in source code once it's working properly.

I think I have a lead on what is going wrong: the issue is not with the code itself, but related to the EnvironmentVariablesRule utility in the tests. The method used by the test utility to install its mock of the ProcessEnvironment is messing-up the internals with JRE on windows. 

I propose to remove this test dependency and restore the original approach, which relied on setting an environment variable in the surefire plugin settings. This has one downside, when you run the test directly from the IDE, you'll need to add the variable to the execution. An assertion in the test validating the presence of that variable could then serve as documentation of this specific aspect. What do you think of this approach @oliverlietz ?
